File: lang/java/cxlib/src/test/java/cx/TableTest.java
package cx;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class TableTest {

    @Test
    public void testFromCxSimple() {
        String src = "[users :table[name age:int]\n" +
                     "  alice 30\n" +
                     "  bob 25\n" +
                     "]";
        Table t = Table.fromCx(src);
        assertEquals(2, t.rowCount());
        assertEquals(2, t.colCount());
    }

    @Test
    public void testFromCxNoTableErrors() {
        IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> Table.fromCx("[product name=alice]")
        );
        assertTrue(e.getMessage().contains("no :table"));
    }

    @Test
    public void testCreateValidatesLen() {
        IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> Table.create(
                Arrays.asList("a", "b"),
                Arrays.asList("int"),
                java.util.Collections.emptyList()
            )
        );
        assertTrue(e.getMessage().contains("len(cols)"));
    }

    @Test
    public void testCreateValidatesUnique() {
        IllegalArgumentException e = assertThrows(
            IllegalArgumentException.class,
            () -> Table.create(
                Arrays.asList("a", "a"),
                Arrays.asList("int", "int"),
                java.util.Collections.emptyList()
            )
        );
        assertTrue(e.getMessage().contains("duplicate"));
    }

    @Test
    public void testRowAndColumn() {
        Table t = Table.create(
            Arrays.asList("a", "b"),
            Arrays.asList("int", "string"),
            Arrays.asList(
                Arrays.asList((Object) 1L, "x"),
                Arrays.asList((Object) 2L, "y")
            )
        );
        Map<String, Object> row = t.row(0);
        assertEquals(1L, row.get("a"));
        assertEquals("x", row.get("b"));
        List<Object> col = t.column("b");
        assertEquals(Arrays.asList("x", "y"), col);
    }

    @Test
    public void testSliceHeadTail() {
        Table t = Table.create(
            Arrays.asList("v"),
            Arrays.asList("int"),
            Arrays.asList(
                Arrays.<Object>asList(1L),
                Arrays.<Object>asList(2L),
                Arrays.<Object>asList(3L),
                Arrays.<Object>asList(4L),
                Arrays.<Object>asList(5L)
            )
        );
        assertEquals(2, t.head(2).rowCount());
        assertEquals(2, t.tail(2).rowCount());
        assertEquals(3, t.slice(1, 4).rowCount());
    }

    @Test
    public void testSelectColsReorders() {
        Table t = Table.create(
            Arrays.asList("a", "b", "c"),
            Arrays.asList("int", "int", "int"),
            Arrays.asList(Arrays.<Object>asList(1L, 2L, 3L))
        );
        Table sel = t.selectCols(Arrays.asList("c", "a"));
        assertEquals(Arrays.asList("c", "a"), sel.cols());
    }

    @Test
    public void testToCx() {
        Table t = Table.create(
            Arrays.asList("a"),
            Arrays.asList("int"),
            Arrays.asList(Arrays.<Object>asList(1L))
        );
        String out = t.toCx();
        assertTrue(out.contains(":table[a:int]"));
    }

    @Test
    public void testToJson() {
        Table t = Table.create(
            Arrays.asList("a"),
            Arrays.asList("int"),
            Arrays.asList(
                Arrays.<Object>asList(1L),
                Arrays.<Object>asList(2L)
            )
        );
        String js = t.toJson();
        assertTrue(js.contains("\"a\":1"));
    }

    @Test
    public void testEquals() {
        Table a = Table.create(
            Arrays.asList("a"),
            Arrays.asList("int"),
            Arrays.asList(Arrays.<Object>asList(1L))
        );
        Table b = Table.create(
            Arrays.asList("a"),
            Arrays.asList("int"),
            Arrays.asList(Arrays.<Object>asList(1L))
        );
        assertEquals(a, b);
    }

    @Test
    public void testFromCxCollectionCells() {
        String src = "[u :table[name tags]\n" +
                     "  alice [admin, user,]\n" +
                     "]";
        Table t = Table.fromCx(src);
        Map<String, Object> row = t.row(0);
        Object tags = row.get("tags");
        assertTrue(tags instanceof List<?>, "tags should be a List, got: " + tags);
    }
}

File: lang/kotlin/cxlib/src/test/kotlin/cx/TableTest.kt
package cx

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class TableTest {

    @Test fun testFromCxSimple() {
        val src = """
            [users :table[name age:int]
              alice 30
              bob 25
            ]
        """.trimIndent()
        val t = Table.fromCx(src)
        assertEquals(2, t.rowCount)
        assertEquals(2, t.colCount)
    }

    @Test fun testFromCxNoTableThrows() {
        val e = assertThrows<IllegalArgumentException> {
            Table.fromCx("[product name=alice]")
        }
        assertTrue(e.message!!.contains("no :table"))
    }

    @Test fun testCreateValidatesLen() {
        val e = assertThrows<IllegalArgumentException> {
            Table.create(listOf("a", "b"), listOf("int"), emptyList())
        }
        assertTrue(e.message!!.contains("len(cols)"))
    }

    @Test fun testCreateValidatesUnique() {
        val e = assertThrows<IllegalArgumentException> {
            Table.create(listOf("a", "a"), listOf("int", "int"), emptyList())
        }
        assertTrue(e.message!!.contains("duplicate"))
    }

    @Test fun testRowAndColumn() {
        val t = Table.create(
            listOf("a", "b"),
            listOf("int", "string"),
            listOf(
                listOf(1L, "x"),
                listOf(2L, "y"),
            )
        )
        val row = t.row(0)
        assertEquals(1L, row["a"])
        assertEquals("x", row["b"])
        assertEquals(listOf<Any?>("x", "y"), t.column("b"))
    }

    @Test fun testSliceHeadTail() {
        val t = Table.create(
            listOf("v"),
            listOf("int"),
            listOf(listOf(1L), listOf(2L), listOf(3L), listOf(4L), listOf(5L)),
        )
        assertEquals(2, t.head(2).rowCount)
        assertEquals(2, t.tail(2).rowCount)
        assertEquals(3, t.slice(1, 4).rowCount)
    }

    @Test fun testSelectColsReorders() {
        val t = Table.create(
            listOf("a", "b", "c"),
            listOf("int", "int", "int"),
            listOf(listOf(1L, 2L, 3L)),
        )
        val sel = t.selectCols(listOf("c", "a"))
        assertEquals(listOf("c", "a"), sel.cols)
    }

    @Test fun testIteration() {
        val t = Table.create(
            listOf("a"),
            listOf("int"),
            listOf(listOf(1L), listOf(2L)),
        )
        var sum = 0L
        for (row in t) sum += row["a"] as Long
        assertEquals(3L, sum)
    }

    @Test fun testToCx() {
        val t = Table.create(
            listOf("a"),
            listOf("int"),
            listOf(listOf(1L)),
        )
        assertTrue(t.toCx().contains(":table[a:int]"))
    }

    @Test fun testToJson() {
        val t = Table.create(
            listOf("a"),
            listOf("int"),
            listOf(listOf(1L), listOf(2L)),
        )
        assertTrue(t.toJson().contains("\"a\":1"))
    }

    @Test fun testEquals() {
        val a = Table.create(listOf("a"), listOf("int"), listOf(listOf(1L)))
        val b = Table.create(listOf("a"), listOf("int"), listOf(listOf(1L)))
        assertEquals(a, b)
    }

    @Test fun testFromCxCollectionCells() {
        val src = """
            [u :table[name tags]
              alice [admin, user,]
            ]
        """.trimIndent()
        val t = Table.fromCx(src)
        val row = t.row(0)
        assertTrue(row["tags"] is List<*>, "tags should be a List, got: ${row["tags"]}")
    }
}
